src/crepe/ComponentManager.cpp

```diff
@@ -21,13 +27,32 @@ void ComponentManager::delete_all_components_of_id(game_object_id_t id) {
 }
 
 void ComponentManager::delete_all_components() {
-	this->components.clear();
+	// Loop through all the ids and delete all components of each id
+	for (game_object_id_t id = 0; id < next_id; id++) {
+		delete_all_components_of_id(id);
+	}
 	this->next_id = 0;
 }
 
 GameObject ComponentManager::new_object(const string & name, const string & tag,
```

Comment:
Ik heb liever dat je de aanpassingen op deze functie weer terugdraait en de `this->next_id = 0;` weg haalt uit `ComponentManager::delete_all_components` (en eventueel andere plekken waar `next_id` gereset wordt.

Reply:
Waar/Hoe wil je dan `this->next_id` op 0 zetten?

Reply:
Niet

Reply:
Dat zou ik wel graag in de code houden, want als we `this->next_id = 0;` niet uitvoeren, wordt de ComponentManager steeds minder efficiënt naarmate de game vordert.

Reply:
Waar moet de component manager *zoeken* op componenten? De lijsten van components blijven misschien groeien wat niet optimaal is voor geheugengebruik (dit is geen requirement dus interesseert mij niet), maar de component manager hoeft toch nergens te zoeken voor componenten? Ik was onder de indruk dat je *of* de complete lijst opvraagt, *of* een specifieke lijst componenten opvraagt met een `game_object_id_t`?

Reply:
Ik wil de `next_id` graag alleen laten optellen omdat het de code veel eenvoudiger maakt. Nu ben je soortvan je eigen memory manager aan het schrijven. Ik weet niet of we eisen moeten gaan stellen aan de snelheid van `add_component` en `delete_components`, maar ik neem aan dat de game programmeur deze functies alleen zal gebruiken bij het aanmaken/opruimen van een scène, en tijdens de scène alleen `active` zal aanpassen. In dat geval hoeven ze ook niet enorm snel te zijn.

Hiernaast valt het mij ook nu pas op dat er geen functie is om één specifiek component te verwijderen van een gameobject. Mij leek het op meerdere momenten al handig als alle *componenten* ook nog een unieke ID zouden hebben, zodat ik bijv. in de ResourceManager eenvoudig bij kan houden voor welke componenten de concrete Resource al aangemaakt is. Ik herinner me ook dat dit bewust niet toegevoegd was als member van de basis Component klasse, maar ben vergeten waarom.

Reply:
> Hiernaast valt het mij ook nu pas op dat er geen functie is om één specifiek component te verwijderen van een gameobject. Mij leek het op meerdere momenten al handig als alle _componenten_ ook nog een unieke ID zouden hebben, zodat ik bijv. in de ResourceManager eenvoudig bij kan houden voor welke componenten de concrete Resource al aangemaakt is. Ik herinner me ook dat dit bewust niet toegevoegd was als member van de basis Component klasse, maar ben vergeten waarom.


Het is wél mogelijk om een specifiek component-type op een specifiek id te deleten. Dus stel dat ik twee Sprites en één Rigidbody aan ID 10 heb gekoppeld, dan is het mogelijk om bijvoorbeeld alleen de twee Sprites te deleten. Het is alleen niet mogelijk om één van de twee Sprites te deleten. Dit zouden we voor nu zo laten, totdat we (tijdens het bouwen van de game) er achter komen dat dit nodig blijkt te zijn.
PS Ik hoop dat (voorlopig) niemand anders, behalve de SceneManager components gaat deleten.

Reply:
Dan vind ik de code nu prima. Omdat je de volgende beschikbare `next_id` wel pas op zoekt bij `add_component` heb ik liever dat je deze variabele hernoemt naar `last_id` omdat de naam niet langer overeenkomt met hoe deze gebruikt wordt.

Reply:
Top! Fijn dat de code dan prima is.

Ik zoek `next_id` volgens mij niet op bij `add_component`?

Reply:
```cpp
	// Find the first available id (taking persistent objects into account)
	while (this->persistent[this->next_id]) {
		this->next_id++;
	}
```
Dit stukje

Reply:
Is inderdaad de GameObject ID oeps

Reply:
Ahaah, daar.
Dit is wel echt de `next_id` (en geen `last_id`), want `next_id` wordt van te voren opgehoogd (precies na het aanmaken van een nieuw gameObject). Dus het is geen `last_id`.